Sources/ShapedArray/Basic.swift

```diff
@@ -49,6 +53,115 @@ extension ShapedArray {
         self.reshaped(to: -1)
     }
 
+    // helper data structure for subarray generation
+    @usableFromInline
+    internal struct SubarrayIndices {
+        @usableFromInline
+        let start: [Int]
+        @usableFromInline
+        let end: [Int]
+    }
+
+    // helper function for subarray indices generation
+    /// takes a shape for the original array, a shape for the subarray, 
+    /// and an axis on which the original array would be divided 
+    /// and returns an array of start and end indices for each subarray
+
+    @usableFromInline
+    internal func _calculateSubarrayIndices(shape: [Int], subarrayShape: [Int], axis: Int) -> [SubarrayIndices] {
```

Comment:
The only use of the`_` prefix for a name is for `public` identifiers such as `public protocol _ShapedArrayProtocol` because they need to be `public` to work yet we want to signal to the user that they shouldn't be used directly. For non-`public` identifiers I think we should omit `_`, in this case that would be `calculateSubarrayIndices`.

However, i think we can improve on the name more too. Possibly
```swift
slicedSubArrayIndices(forShape shape: [Int], subarrayShape: [Int], alongAxis axis: Int) -> [SubarrayIndices]
```
?

Lastly, and maybe this isn't possible with `@usableFromInline`, but if this is only used in 1 function I usually move the function declaration inside of the function that uses it. It adds some indentation but makes it clear at point of use that it isn't used elsewhere. Up to you though, nbd either way.

Reply:
or maybe `splitSubArrayIndices` to match the calling function..